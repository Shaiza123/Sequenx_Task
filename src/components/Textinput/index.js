import { View, Text, Keyboard } from 'react-native'
import React from 'react'
import { TextInput } from 'react-native-paper';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons'
import FontAwesome from 'react-native-vector-icons/FontAwesome'
import styles from './style';
import { heightPercentageToDP as hp, widthPercentageToDP as wp } from 'react-native-responsive-screen';
import { Formik } from 'formik';
import { Button } from 'react-native-paper'
import { Picker } from '@react-native-picker/picker';
import PropTypes from 'prop-types';

const Textinput = (props) => {
    return (
        <View>
            <Formik
                initialValues={{ name: '', price: '', material: '' }}
                validationSchema={props?.productFormSchema}
                onSubmit={(values, { resetForm }) => {
                    props?.addProduct(values?.name, values?.price, values?.material, resetForm);
                }}
            >
                {({ handleChange, handleBlur, handleSubmit, values, errors, touched }) => (
                    <>
                        <TextInput
                            placeholder="Product Name"
                            onChangeText={handleChange('name')}
                            onBlur={handleBlur('name')}
                            value={values.name}
                            ref={props?.productNameRef}
                            onSubmitEditing={() => props?.productPriceRef.current.focus()}
                            style={styles.input}
                            autoFocus={false}
                            underlineColor="#0147AB"
                            activeUnderlineColor='#0147AB'
                            textColor='#707070'
                            mode='flat'
                            keyboardType="default"
                            left={<TextInput.Icon icon={() => <FontAwesome name="user" size={hp(2.5)} color={'#000'} />} />}
                        />
                        {touched.name && errors.name && <Text style={styles.errorText}>{errors.name}</Text>}

                        <TextInput
                            placeholder="Price"
                            onChangeText={handleChange('price')}
                            onBlur={handleBlur('price')}
                            value={values.price}
                            ref={props?.productPriceRef}
                            onSubmitEditing={() => Keyboard.dismiss()}
                            style={styles.input}
                            autoFocus={false}
                            underlineColor="#0147AB"
                            activeUnderlineColor='#0147AB'
                            textColor='#707070'
                            mode='flat'
                            keyboardType="numeric"
                            left={<TextInput.Icon icon={() => <MaterialCommunityIcons name="currency-usd" size={hp(2.5)} color={'#000'} />} />}
                        />
                        {touched.price && errors.price && <Text style={styles.errorText}>{errors.price}</Text>}
                        <Picker
                            selectedValue={values.material}
                            style={styles.picker}
                            onValueChange={(itemValue) =>
                                handleChange('material')(itemValue)
                            }>
                            <Picker.Item label="Select Material" value="" />
                            <Picker.Item label="Metal" value="Metal" />
                            <Picker.Item label="Plastic" value="Plastic" />
                            <Picker.Item label="Bronze" value="Bronze" />
                            <Picker.Item label="Wooden" value="Wooden" />
                            <Picker.Item label="Frozen" value="Frozen" />
                            <Picker.Item label="Cotton" value="Cotton" />
                            <Picker.Item label="Rubber" value="Rubber" />
                            <Picker.Item label="Soft" value="Soft" />
                            <Picker.Item label="Concrete" value="Concrete" />
                            <Picker.Item label="Granite" value="Granite" />
                        </Picker>
                        {touched.material && errors.material && <Text style={styles.errorText}>{errors.material}</Text>}
                        <Button style={styles.button} textColor='#fff' mode='contained' buttonColor={'#fa6e20'}
                            disabled={values.name && values.price && values.material ? false : true}
                            contentStyle={{
                                paddingHorizontal: wp(15),
                                paddingVertical: wp(2)
                            }}
                            onPress={handleSubmit}
                        >
                            <Text style={styles.buttonText}>Add Product</Text>
                        </Button>
                    </>
                )}
            </Formik>
        </View>

    )
}

Textinput.propTypes = {
    productFormSchema: PropTypes.object.isRequired,
    addProduct: PropTypes.func.isRequired,
    productNameRef: PropTypes.object.isRequired,
    productPriceRef: PropTypes.object.isRequired,
};

export default Textinput