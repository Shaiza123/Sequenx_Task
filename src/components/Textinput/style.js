import { StyleSheet } from "react-native";
import { RFPercentage } from "react-native-responsive-fontsize";
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';

const COLORS = {
    white: '#fff',
    black: '#000',
    backgroundColor: 'transparent',
    borderColor: '#ccc',
    error: '#ff0000',
};

const styles = StyleSheet.create({
    button: {
        alignItems: 'center',
        borderRadius: hp('5%'),
        marginTop: hp('2%'),
        width: '100%',
    },
    buttonText: {
        color: COLORS.white,
        fontSize: RFPercentage(2),
        fontWeight: 'bold',
    },
    errorText: {
        color: COLORS.error,
        fontSize: RFPercentage(1.8),
        marginBottom: hp('1%'),
    },
    input: {
        backgroundColor:COLORS.backgroundColor,
        fontSize: RFPercentage(2),
        marginBottom: hp('2%'),
        paddingHorizontal: wp('4%'),
        width: '100%',
    },
    label: {
        fontSize: RFPercentage(2),
        fontWeight: 'bold',
        marginBottom: hp('1%'),
    },
    picker: {
        borderColor: COLORS.borderColor,
        borderRadius: hp('1%'),
        borderWidth: 1,
        marginBottom: hp('2%'),
        width: '100%',
    },
  });

export default styles;