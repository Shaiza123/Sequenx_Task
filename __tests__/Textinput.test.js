import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import Textinput from '../src/components/Textinput';

// Mock dependencies
const mockProductFormSchema = {
  validate: jest.fn(() => Promise.resolve({})),
  validateSync: jest.fn(() => ({})),
};

const mockAddProduct = jest.fn();

const mockProductNameRef = {
  current: { focus: jest.fn() },
};

const mockProductPriceRef = {
  current: { focus: jest.fn() },
};
jest.useFakeTimers();

test('render correctly', () => {
  const { toJSON } = render(
    <Textinput
      productFormSchema={mockProductFormSchema}
      addProduct={mockAddProduct}
      productNameRef={mockProductNameRef}
      productPriceRef={mockProductPriceRef}
    />
  );

  expect(toJSON()).toMatchSnapshot();
});

test('renders correctly and triggers form submission', async () => {
  const { getByPlaceholderText, getByText, getByTestId } = render(
    <Textinput
      productFormSchema={mockProductFormSchema}
      addProduct={mockAddProduct}
      productNameRef={mockProductNameRef}
      productPriceRef={mockProductPriceRef}
    />
  );

  const nameInput = getByPlaceholderText('Product Name');
  const priceInput = getByPlaceholderText('Price');
  const materialPicker = getByTestId('materialPicker');
  const addButton = getByText('Add Product');

  fireEvent.changeText(nameInput, 'Product 1');
  fireEvent.changeText(priceInput, '10');
  fireEvent(materialPicker, 'valueChange', 'Metal');
  fireEvent.press(addButton);
  

  await waitFor(async() => {
    expect(mockAddProduct).toHaveBeenCalledWith('Product 1', '10', 'Metal', expect.any(Function));
  });

  jest.runOnlyPendingTimers();
});


afterAll(() => {
  jest.useRealTimers();
});