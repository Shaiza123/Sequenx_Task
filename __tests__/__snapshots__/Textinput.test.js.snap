// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "backgroundColor": "transparent",
          "marginBottom": 26.5,
          "width": "100%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#0147AB",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 56,
          },
        ]
      }
    >
      <TextInput
        autoFocus={false}
        cursorColor="#0147AB"
        editable={true}
        keyboardType="default"
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Product Name"
        placeholderTextColor="rgba(73, 69, 79, 1)"
        selectionColor="#0147AB"
        style={
          [
            {
              "margin": 0,
            },
            {
              "height": 56,
            },
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "color": "#707070",
              "fontFamily": "System",
              "fontSize": 27,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": 32.4,
              "minWidth": 65,
              "paddingLeft": 30,
              "paddingRight": 30,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            {
              "marginLeft": 54,
              "paddingLeft": 16,
            },
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        [
          {
            "alignItems": "center",
            "height": 24,
            "justifyContent": "center",
            "position": "absolute",
            "width": 24,
          },
          {
            "left": 16,
            "top": 16,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 20,
            "height": 40,
            "margin": 0,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": 40,
          }
        }
        testID="left-icon-adornment-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 20,
              "borderWidth": 0,
              "elevation": 0,
              "flex": 1,
              "overflow": "hidden",
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="left-icon-adornment-container"
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityTraits="button"
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            centered={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                [
                  {
                    "alignItems": "center",
                    "flexGrow": 1,
                    "justifyContent": "center",
                  },
                  {
                    "borderRadius": 20,
                  },
                ],
              ]
            }
            testID="left-icon-adornment"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#000",
                    "fontSize": 33.5,
                  },
                  undefined,
                  {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "backgroundColor": "transparent",
          "marginBottom": 26.5,
          "width": "100%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#0147AB",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 56,
          },
        ]
      }
    >
      <TextInput
        autoFocus={false}
        cursorColor="#0147AB"
        editable={true}
        keyboardType="numeric"
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Price"
        placeholderTextColor="rgba(73, 69, 79, 1)"
        selectionColor="#0147AB"
        style={
          [
            {
              "margin": 0,
            },
            {
              "height": 56,
            },
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "color": "#707070",
              "fontFamily": "System",
              "fontSize": 27,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": 32.4,
              "minWidth": 65,
              "paddingLeft": 30,
              "paddingRight": 30,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            {
              "marginLeft": 54,
              "paddingLeft": 16,
            },
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        [
          {
            "alignItems": "center",
            "height": 24,
            "justifyContent": "center",
            "position": "absolute",
            "width": 24,
          },
          {
            "left": 16,
            "top": 16,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 20,
            "height": 40,
            "margin": 0,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": 40,
          }
        }
        testID="left-icon-adornment-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 20,
              "borderWidth": 0,
              "elevation": 0,
              "flex": 1,
              "overflow": "hidden",
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="left-icon-adornment-container"
        >
          <View
            accessibilityComponentType="button"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityTraits="button"
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            centered={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 6,
                "left": 6,
                "right": 6,
                "top": 6,
              }
            }
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                [
                  {
                    "alignItems": "center",
                    "flexGrow": 1,
                    "justifyContent": "center",
                  },
                  {
                    "borderRadius": 20,
                  },
                ],
              ]
            }
            testID="left-icon-adornment"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#000",
                    "fontSize": 33.5,
                  },
                  undefined,
                  {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              󰇁
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 13.5,
        "borderWidth": 1,
        "marginBottom": 26.5,
        "width": "100%",
      }
    }
  >
    <RNCPicker
      items={
        [
          {
            "label": "Select Material",
            "testID": undefined,
            "textColor": undefined,
            "value": "",
          },
          {
            "label": "Metal",
            "testID": undefined,
            "textColor": undefined,
            "value": "Metal",
          },
          {
            "label": "Plastic",
            "testID": undefined,
            "textColor": undefined,
            "value": "Plastic",
          },
          {
            "label": "Bronze",
            "testID": undefined,
            "textColor": undefined,
            "value": "Bronze",
          },
          {
            "label": "Wooden",
            "testID": undefined,
            "textColor": undefined,
            "value": "Wooden",
          },
          {
            "label": "Frozen",
            "testID": undefined,
            "textColor": undefined,
            "value": "Frozen",
          },
          {
            "label": "Cotton",
            "testID": undefined,
            "textColor": undefined,
            "value": "Cotton",
          },
          {
            "label": "Rubber",
            "testID": undefined,
            "textColor": undefined,
            "value": "Rubber",
          },
          {
            "label": "Soft",
            "testID": undefined,
            "textColor": undefined,
            "value": "Soft",
          },
          {
            "label": "Concrete",
            "testID": undefined,
            "textColor": undefined,
            "value": "Concrete",
          },
          {
            "label": "Granite",
            "testID": undefined,
            "textColor": undefined,
            "value": "Granite",
          },
        ]
      }
      numberOfLines={1}
      onChange={[Function]}
      selectedIndex={0}
      style={
        [
          {
            "height": 216,
          },
          undefined,
        ]
      }
      testID="materialPicker"
    />
  </View>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(28, 27, 31, 0.12)",
        "borderRadius": 66.5,
        "marginTop": 26.5,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "width": "100%",
      }
    }
    testID="button-container-outer-layer"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(28, 27, 31, 0.12)",
          "borderColor": "transparent",
          "borderRadius": 66.5,
          "borderStyle": "solid",
          "borderWidth": 0,
          "flex": undefined,
          "minWidth": 64,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 66.5,
            },
          ]
        }
        testID="button"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "paddingHorizontal": 112.5,
                "paddingVertical": 15,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 24,
                      "marginVertical": 10,
                    },
                    undefined,
                    false,
                    {
                      "color": "rgba(28, 27, 31, 0.38)",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    undefined,
                  ],
                ],
              ]
            }
            testID="button-text"
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 27,
                  "fontWeight": "bold",
                }
              }
            >
              Add Product
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
